### ISSUE TITLE
[Proposal] We need to integrate CLA check component

### ISSUE THREAD

**Issue by WuShaoling** [hypertrons/hypertrons#299]
Contributors sign CLA agreement is crucial in open source collaboration. We are going to integrate the features of cross-platform CLA check into Hypertrons:

1. For the GitHub platform, we can refer to [cla-assistant](https://cla-assistant.io/).
2. For the Gitee and GitLab platform, we can implement it by listening to PR events. After receiving the PR event, we need to check whether the submitter of the PR has signed the CLA agreement. If it has been signed, we need to add a label to indicate that the PR can be merged. If not, add a comment with a link, committers can click the link to jump to the signing page, then we need to save user's information to the database.
3. The above process brings two new requirements:
   3.1  We need a user system to identify each committer. We can use the user's email to achieve cross-platform authentication. There will be a special case, that is, the contributor's GitHub account and Gitee account do not use the same email, in this case, the committer must sign the agreement twice.
   3.2  We need a database to store user's information.

**Comment by will-ww** [comment]
Great !

**Comment by WillemJiang** [comment]
Using the email address can help us identify if the PR submitter signed the CLA or not.  The challenge part is we need additional information to connect PR author with their email addresses. ( I cannot find the API from gitee at this time). We may need to verify the email address if we let user to submit the email address themself through the web page.

**Comment by frank-zsy** [comment]
Even on GitHub, we can not identify users by email because email may not be public for every account, however there will always be an email address relate to a certain commit although it is not usable if users make change directly from web page.

**Comment by WillemJiang** [comment]
For the CLA case we just need to verify the median or large change and ignore the minor change which is done through the github web page. If I remember right, Apache only ask the contributor to submit the CLA if patch size is big enough. Not every patch need a CLA.
BTW, There are some check service to block the change which is not signed by a valid email address.

**Comment by frank-zsy** [comment]
OK, so I think we can use commit email address to identify users which also requires users using same email address across different platform, if we can relate user accounts across platforms that will be better.

**Comment by freesky-edward** [comment]
this is an interesting case I think that would be a single part of the whole open source infrastructure. @TommyLike

**Comment by TommyLike** [comment]
Are we going to build a band new CLA cross-platform system or just utilize and integrate some exisiting system , such as CLA assistant? @WuShaoling

**Comment by TommyLike** [comment]
What I can tell from the proposal is we gonna have two individual systems that are serving for github and gitee, what if we have gitlab or gitxx in the future? since the systems are isolated, how about adding some features in project [cla-assistant](https://github.com/cla-assistant/cla-assistant) to have it support different systems?

**Comment by freesky-edward** [comment]
the email should not be the publish email but be the commit email address which supported by both on github and gitee,  it is the one which will configuration by contributor when `git commit`. 
so that the checker should base on each committer's commit email of one PR instead of PR author's public email. please consider one senario: there are more than two contributors cooperated on one feature, they will commit their codes into one branch and create the PR from that branch. if any one of them didn't sign the CLA, the bot should give the sign notification message and mark this PR as `cla/no`.

**Comment by frank-zsy** [comment]
But there is a tricky thing that how can we be sure the user have the access to an email address? In CLA assistant users need to login via GitHub OAuth2 service so that the system knows the user ***does own*** the account, but if we use email address to sign CLA, how can we know the user ***does own*** the email account?

**Comment by WillemJiang** [comment]
@frank-zsy  I think we can let use fill the email account themself on the CLA sign page and add a process to verify the email address if we want.

**Comment by frank-zsy** [comment]
Yes, and I think we don't even need to let the user fill the email, if the robot find any author of the commit didn't sign the CLA, it can send an email to the account with the CLA content and the user can reply via the account to sign CLA, so it's quite automatic and totally platform independent, and also can deal with multiple author situation which CLA assistant may not consider right now.

**Comment by WillemJiang** [comment]
It's good idea to send an email to remind the user to sign the CLA if he doesn't sign.  But I think it could be handy if there is a check on the PR with detail instruction to fix the issue, just like [dco](https://probot.github.io/apps/dco/) does.

**Comment by WillemJiang** [comment]
> What I can tell from the proposal is we gonna have two individual systems that are serving for github and gitee, what if we have gitlab or gitxx in the future? since the systems are isolated, how about adding some features in project [cla-assistant](https://github.com/cla-assistant/cla-assistant) to have it support different systems?

Now we are talking about adding the CLA check feature as an extend component of hypertrons. We may think a more generic way to provide this kind of service across the different code hosting platform, otherwise it doesn't save our time if we need to deal with each platform one by one.

**Comment by frank-zsy** [comment]
> It's good idea to send an email to remind the user to sign the CLA if he doesn't sign. But I think it could be handy if there is a check on the PR with detail instruction to fix the issue, just like [dco](https://probot.github.io/apps/dco/) does.

Yes, we can add a check with detailed information about the email and description of the component.
